rasa/utils/tensorflow/layers_utils.py

```diff
@@ -71,14 +71,24 @@ def get_candidate_values(
     return candidate_values  # (batch_size, num_candidates, tf.shape(x)[-1])
 
 
-def reduce_mean_equal(x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+def reduce_mean_equal(
+    x: tf.Tensor, y: tf.Tensor, mask: Optional[tf.Tensor] = None
+) -> tf.Tensor:
     """Computes the mean number of matches between x and y.
 
     Args:
-        x: Any numeric tensor
-        y: Another tensor with same shape and type as x
+        x: Any numeric tensor.
+        y: Another tensor with same shape and type as x.
+        mask: Tensor with a mask to distinguish actual indices from padding indices.
+            Shape should be the same as `x` and `y`.
 
     Returns:
         The mean of "x == y"
     """
-    return tf.reduce_mean(tf.cast(tf.math.equal(x, y), tf.float32))
+    if mask is None:
+        return tf.reduce_mean(tf.cast(tf.math.equal(x, y), tf.float32))
+    else:
+        equal_indices = tf.cast(tf.math.equal(x, y), tf.float32) * mask
+        return tf.reduce_mean(
+            tf.reduce_sum(equal_indices, axis=-1) / tf.reduce_sum(mask, axis=-1)
```

Comment:
Why is this only reduced along the last axis when a mask is passed?

Reply:
Hmm, good question. There is an assumption that the actual mask values are on the last dimension and hence the first `reduce_mean` is needed only on that axis. The next `reduce_mean` is across all axis.